### Question 1 :-  A company wants to estimate the average time spent by customers on their website per session. They collect a random sample of 100 customers and find that the sample mean time spent is 4.5 minutes with a sample standard deviation of 1.2 minutes. Calculate a 99% confidence interval for the true population mean time spent on the website per session.

In [1]:
import scipy.stats as stats
import math

# sample information
n = 100
sample_mean = 4.5
sample_std = 1.2

# calculate the standard error of the mean
se = sample_std / math.sqrt(n)

# calculate the critical value for a 99% confidence level
alpha = 0.01  # 1 - confidence level
cv = stats.norm.ppf(1 - alpha/2)

# calculate the confidence interval
lower_ci = sample_mean - cv * se
upper_ci = sample_mean + cv * se

# print the result
print("The 99% confidence interval for the true population mean time spent on the website per session is: ({:.3f}, {:.3f}) minutes".format(lower_ci, upper_ci))


The 99% confidence interval for the true population mean time spent on the website per session is: (4.191, 4.809) minutes


### Question 2: A car manufacturer is interested in estimating the mean gas mileage of their new SUV. A sample of 25 SUVs is taken, and their mean gas mileage is found to be 28.6 miles per gallon with a standard deviation of 2.8 miles per gallon. Calculate a 95% confidence interval for the true mean gas mileage of the SUV.

In [2]:
import scipy.stats as stats
import math

# sample information
n = 25
sample_mean = 28.6
sample_std = 2.8

# calculate the standard error of the mean
se = sample_std / math.sqrt(n)

# calculate the critical value for a 95% confidence level
alpha = 0.05  # 1 - confidence level
df = n - 1  # degrees of freedom
cv = stats.t.ppf(1 - alpha/2, df)

# calculate the confidence interval
lower_ci = sample_mean - cv * se
upper_ci = sample_mean + cv * se

# print the result
print("The 95% confidence interval for the true mean gas mileage of the SUV is: ({:.4f}, {:.4f}) miles per gallon".format(lower_ci, upper_ci))


The 95% confidence interval for the true mean gas mileage of the SUV is: (27.4442, 29.7558) miles per gallon


In [4]:
import pandas as pd

In [6]:
ind_temp = pd.read_csv("ind_temp.csv")
ind_temp

dt  AverageTemperature  AverageTemperatureUncertainty Country
0     1796-01-01              17.044                          2.044   India
1     1796-02-01              19.193                          1.359   India
2     1796-03-01              22.319                          2.125   India
3     1796-04-01              27.233                          1.510   India
4     1796-05-01              30.035                          1.338   India
...          ...                 ...                            ...     ...
2608  2013-05-01              31.014                          0.205   India
2609  2013-06-01              28.766                          0.299   India
2610  2013-07-01              27.012                          0.197   India
2611  2013-08-01              26.555                          0.242   India
2612  2013-09-01                 NaN                            NaN   India

[2613 rows x 4 columns]

`Question -3:` Dataset - India temperature dataset.

Data link: https://drive.google.com/file/d/1whb8Rjdk9TBKOwVi8NwZD1Tc8Tdv7iEc/view?usp=share_link

Kaggle Link : (Just take IND data for this question, have to apply filtering). https://www.kaggle.com/datasets/berkeleyearth/climate-change-earth-surface-temperature-data

A climate research organization wants to estimate the average temperature of a certain country. They collect temperature data for 2613 days but due to certain limitations, they only have information about the average temperature for 2508 days. The organization assumes that the population follows a normal distribution and wants to estimate the population mean temperature with a 95% confidence level.

* Standard deviation is assumed as given data standard deviation. z Procedure
* Standard deviation is not given. Apply t Procedure

In [7]:
#Standard deviation is assumed as given data standard deviation. z Procedure
import pandas as pd
import numpy as np
import scipy.stats as stats

# load the data
ind_temp = pd.read_csv("ind_temp.csv")

# calculate the sample mean temperature and population standard deviation
sample_mean = ind_temp["AverageTemperature"].mean()
pop_std = ind_temp["AverageTemperature"].std()

# sample information
n = 2508

# calculate the standard error of the mean
se = pop_std / np.sqrt(n)

# calculate the critical value for a 95% confidence level
alpha = 0.05  # 1 - confidence level
cv = stats.norm.ppf(1 - alpha/2)

# calculate the confidence interval
lower_ci = sample_mean - cv * se
upper_ci = sample_mean + cv * se

# print the result
print("The 95% confidence interval for the population mean temperature is: ({:.2f}, {:.2f}) degrees Celsius".format(lower_ci, upper_ci))


The 95% confidence interval for the population mean temperature is: (23.70, 24.04) degrees Celsius


In [8]:
#Standard deviation is not given. Apply t Procedure
import pandas as pd
import numpy as np
import scipy.stats as stats

# load the data
ind_temp = pd.read_csv("ind_temp.csv")

# calculate the sample mean temperature and sample standard deviation
sample_mean = ind_temp["AverageTemperature"].mean()
sample_std = ind_temp["AverageTemperature"].std(ddof=1)  # use n-1 degrees of freedom

# sample information
n = 2508

# calculate the standard error of the mean
se = sample_std / np.sqrt(n)

# calculate the critical value for a 95% confidence level
alpha = 0.05  # 1 - confidence level
df = n - 1  # degrees of freedom
cv = stats.t.ppf(1 - alpha/2, df)

# calculate the confidence interval
lower_ci = sample_mean - cv * se
upper_ci = sample_mean + cv * se

# print the result
print("The 95% confidence interval for the population mean temperature is: ({:.2f}, {:.2f}) degrees Celsius".format(lower_ci, upper_ci))


The 95% confidence interval for the population mean temperature is: (23.70, 24.04) degrees Celsius


`Question 5:` BMW Dataset : https://drive.google.com/file/d/18amVv1WKKHEmwyvVy0RuXp4A6xrGOelc/view?usp=share_link



* Task 1:
The sales manager of a used car company wants to know what is the average selling price of all the used bmw cars.
The analyst can collect only a sample of sales of 500 cars in the area. Since this estimate is going to be used by the company to strategize sales of his company, the sample mean should be a good approximation of all the account.
What level of confidence is the sales manager going to be satisfied with? What +/- interval number is going to be acceptable?

* Task 2:
In addition to the price of the car, the manager also wants to now know the average mileage that the car has been driven. But the manager does not have the population standard deviation for the mileage. But the mileage data is available only from the 25 cars that they have sold so far.
How do the analyst approach this problem to calculate 95% confidence interval, with only 25 samples?

* Task 3:
The manager is not happy with both the intervals (intervals from task 1 and 2) as the interval for the confidence is very high. The manager now asks the analyst to estimate the average price of the car (similar to task 1) but within a bound of 750 from the mean with 95% condidence level. How many sample does the analyst have to collect to arrive at this confidence interval level?

* Task 4:
Conversely, after the analyst started collecting the data, after a week he was only able to collect for 420 samples, though he needs 540 samples for Bound of 1,000, if he has to do the analysis now, what is the best interval he can acheive for 95% confidence levels?

In [10]:
bmw = pd.read_csv("bmw.csv")
bmw

model  year  price transmission  mileage fuelType  tax   mpg  \
0       5 Series  2014  11200    Automatic    67068   Diesel  125  57.6   
1       6 Series  2018  27000    Automatic    14827   Petrol  145  42.8   
2       5 Series  2016  16000    Automatic    62794   Diesel  160  51.4   
3       1 Series  2017  12750    Automatic    26676   Diesel  145  72.4   
4       7 Series  2014  14500    Automatic    39554   Diesel  160  50.4   
...          ...   ...    ...          ...      ...      ...  ...   ...   
10776         X3  2016  19000    Automatic    40818   Diesel  150  54.3   
10777   5 Series  2016  14600    Automatic    42947   Diesel  125  60.1   
10778   3 Series  2017  13100       Manual    25468   Petrol  200  42.8   
10779   1 Series  2014   9930    Automatic    45000   Diesel   30  64.2   
10780         X1  2017  15981    Automatic    59432   Diesel  125  57.6   

       engineSize  
0             2.0  
1             2.0  
2             3.0  
3             1.5  
4             3.0  
...           ...  
10776         2.0  
10777         2.0  
10778         2.0  
10779         2.0  
10780         2.0  

[10781 rows x 9 columns]

### Task 1: The sales manager of a used car company wants to know what is the average selling price of all the used bmw cars. The analyst can collect only a sample of sales of 500 cars in the area. Since this estimate is going to be used by the company to strategize sales of his company, the sample mean should be a good approximation of all the account. What level of confidence is the sales manager going to be satisfied with? What +/- interval number is going to be acceptable?

In [11]:
import pandas as pd
import numpy as np
import scipy.stats as stats

# load the data
bmw = pd.read_csv("bmw.csv")

# calculate the sample mean selling price and sample standard deviation
sample_mean = bmw["price"].mean()
sample_std = bmw["price"].std(ddof=1)  # use n-1 degrees of freedom

# sample information
n = 500

# calculate the standard error of the mean
se = sample_std / np.sqrt(n)

# calculate the critical value and confidence interval for different levels of confidence
df = n - 1  # degrees of freedom
conf_levels = [0.90, 0.95, 0.99]
for conf_level in conf_levels:
    alpha = 1 - conf_level
    cv = stats.t.ppf(1 - alpha/2, df)
    lower_ci = sample_mean - cv * se
    upper_ci = sample_mean + cv * se
    interval = cv * se
    print("At a {:.0%} level of confidence, the 95% confidence interval for the population mean selling price is: ({:.2f}, {:.2f}) with a margin of error of +/- {:.2f}".format(conf_level, lower_ci, upper_ci, interval))


At a 90% level of confidence, the 95% confidence interval for the population mean selling price is: (21892.12, 23574.70) with a margin of error of +/- 841.29
At a 95% level of confidence, the 95% confidence interval for the population mean selling price is: (21730.38, 23736.44) with a margin of error of +/- 1003.03
At a 99% level of confidence, the 95% confidence interval for the population mean selling price is: (21413.35, 24053.46) with a margin of error of +/- 1320.06


### Task 2: In addition to the price of the car, the manager also wants to now know the average mileage that the car has been driven. But the manager does not have the population standard deviation for the mileage. But the mileage data is available only from the 25 cars that they have sold so far. How do the analyst approach this problem to calculate 95% confidence interval, with only 25 samples?

In [14]:
import pandas as pd
import numpy as np
from scipy.stats import t

# Load the data
bmw = pd.read_csv('bmw.csv')

# Calculate the sample mean and standard deviation for mileage
mileage_mean = bmw['mileage'].mean()
mileage_std = bmw['mileage'].std(ddof=1)

# Calculate the SEM(Standard Error of the Mean)
#df=n-1
n = 25
sem = mileage_std / np.sqrt(n)

# Calculate the t-value for a 95% confidence level with 24 degrees of freedom
#1 - alpha/2 =0.975
t_value = t.ppf(0.975, df=24)

# Calculate the Margin of Error(MOE)
moe = t_value * sem

# Calculate the lower and upper bounds of the confidence interval
lower_bound = mileage_mean - moe
upper_bound = mileage_mean + moe

# Print the results
print("The 95% confidence interval for the average mileage is: ({:.2f}, {:.2f})".format(lower_bound, upper_bound))


The 95% confidence interval for the average mileage is: (15118.39, 35875.59)


### Task 3: The manager is not happy with both the intervals (intervals from task 1 and 2) as the interval for the confidence is very high. The manager now asks the analyst to estimate the average price of the car (similar to task 1) but within a bound of 750 from the mean with 95% confidence level. How many sample does the analyst have to collect to arrive at this confidence interval level?

In [17]:
import pandas as pd
import numpy as np
from scipy.stats import t

# Load the BMW dataset
bmw = pd.read_csv('bmw.csv')

# Calculate the sample standard deviation of the price
std_dev = np.std(bmw['price'], ddof=1)

# Set the desired margin of error and the confidence level
margin_error = 750
conf_level = 0.95

# Calculate the z-score corresponding to the confidence level
z = t.ppf((1 + conf_level) / 2, len(bmw)-1)

# Calculate the required sample size
n = ((z * std_dev) / margin_error) ** 2

print(f"The required sample size is {int(n)}.")


The required sample size is 890.


### Task 4: Conversely, after the analyst started collecting the data, after a week he was only able to collect for 420 samples, though he needs 540 samples for Bound of 1,000, if he has to do the analysis now, what is the best interval he can acheive for 95% confidence levels?

In [20]:
import pandas as pd
import numpy as np
from scipy import stats

# Load the BMW dataset
bmw = pd.read_csv('bmw.csv')

# Select a sample of 420 cars
sample = bmw.sample(n=420, random_state=1)

# Calculate the sample mean and standard deviation of the price
sample_mean = np.mean(sample['price'])
sample_std_dev = np.std(sample['price'], ddof=1)

# Set the desired confidence level
conf_level = 0.95

# Calculate the standard error of the mean
sem = sample_std_dev / np.sqrt(420)

# Calculate the t-value for a two-tailed test
t_value = stats.t.ppf(1 - (1 - conf_level) / 2, df=419)

# Calculate the margin of error
moe = t_value * sem

# Calculate the lower and upper bounds of the confidence interval
lower_bound = sample_mean - moe
upper_bound = sample_mean + moe

print(f"The 95% confidence interval is [{lower_bound}, {upper_bound}].")


The 95% confidence interval is [22336.805755833844, 24523.04662511854].
